packages/client/hmi-client/src/components/widgets/VegaChart.vue

```diff
@@ -21,15 +21,39 @@
 </template>
 
 <script setup lang="ts">
-import embed, { Result, VisualizationSpec } from 'vega-embed';
-import { Config as VgConfig } from 'vega';
-import { Config as VlConfig } from 'vega-lite';
+import { format } from 'd3';
+import embed, { Config, Result, VisualizationSpec } from 'vega-embed';
 import Button from 'primevue/button';
 import Dialog from 'primevue/dialog';
-
+import { countDigits, fixPrecisionError } from '@/utils/number';
 import { ref, watch, toRaw, isRef, isReactive, isProxy, computed, h, render } from 'vue';
 
-export type Config = VgConfig | VlConfig;
+const NUMBER_FORMAT = '.3~s';
+
+// Define the custom expression functions that can be registered and used in the Vega charts
+const expressionFunctions = {
+	// chartNumberFormatter is a custom number format that will display numbers in a more readable format
+	chartNumberFormatter: (value: number) => {
+		const correctedValue = fixPrecisionError(value);
+		if (value > -1 && value < 1) {
+			return countDigits(correctedValue) > 6 ? correctedValue.toExponential(3) : correctedValue.toString();
+		}
+		return format(NUMBER_FORMAT)(correctedValue);
```

Comment:
Why not using the number util? You spent all that time improving it, and not using it at the end.

Reply:
I think existing `displayMumber` util isn't the best fit for display number for the charts especially for the axis since it either display long numbers as exponential or numbers with spaces. 

For numbers > 1 or < -1, I think it's better to stick with current d3 formatter for the charts (e.g we want 100K instead of 1 00 000 or 1M instead of 1.000e+6). 

For number like `0.00001`, I think it's better to display as it is instead of `0.000 01`  that you would get with the existing function. 

Also existing function didn't handle the precession error, and you would get axiss like 0.1 0.2 '3.000e-1' 0.4, without correcting precision error. 

So I implemented some enhancements to the existing chart and also used custom function for the charts for better numbers.

Reply:
Thank you for the detailled answer!